src/main/java/com/arpnetworking/metrics/common/sources/TcpLineSource.java

```diff
@@ -0,0 +1,245 @@
+            return getContext().actorOf(Props.create(
+                    TcpRequestHandlerActor.class,
+                    getSink(),
+                    connected.remoteAddress()));
+        }
+
+        /**
+         * Constructor.
+         *
+         * @param source The {@link TcpLineSource} to send notifications through.
+         */
+        /* package private */ TcpListenerActor(final TcpLineSource source) {
+            super(source);
+        }
+    }
+
+    /**
+     * Internal actor to processRecords requests.
+     */
+    /* package private */ static final class TcpRequestHandlerActor extends UntypedActor {
+
+        /* package private */ TcpRequestHandlerActor(
+                final TcpLineSource sink,
+                final InetSocketAddress remoteAddress) {
+            _sink = sink;
+            _remoteAddress = remoteAddress;
+        }
+
+        @Override
+        public void onReceive(final Object message) throws Throwable {
+            if (message instanceof Tcp.Received) {
+                final ByteString data = ((Tcp.Received) message).data();
+
+                LOGGER.trace()
+                        .setMessage("Tcp data received")
+                        .addData("name", _sink.getName())
+                        .addData("remoteAddress", _remoteAddress.getAddress().getHostAddress())
+                        .addData("remotePort", _remoteAddress.getPort())
+                        .addData("data", data)
+                        .log();
+
+                try {
+                    processData(data);
+                    // CHECKSTYLE.OFF: IllegalCatch - Ensure all exceptions are logged (this is top level)
+                } catch (final RuntimeException e) {
+                    // CHECKSTYLE.ON: IllegalCatch
+                    BAD_REQUEST_LOGGER.warn()
+                            .setMessage("Error processing data")
+                            .addData("name", _sink.getName())
+                            .addData("remoteAddress", _remoteAddress.getAddress().getHostAddress())
+                            .addData("remotePort", _remoteAddress.getPort())
+                            .addData("data", data)
+                            .setThrowable(e)
+                            .log();
+                }
+            } else if (message instanceof Tcp.ConnectionClosed) {
+                getContext().stop(getSelf());
+                LOGGER.debug()
+                        .setMessage("Tcp connection close")
+                        .addData("name", _sink.getName())
+                        .addData("remoteAddress", _remoteAddress.getAddress().getHostAddress())
+                        .addData("remotePort", _remoteAddress.getPort())
+                        .log();
+            }
+        }
+
+        private void processData(final ByteString data) {
+            // Process datagrams line by line buffering any partial lines
+            int indexStart = 0;
+            int indexEnd = data.indexOf('\n');
```

Comment:
Your Graphite protocol parser (at least) handles multiple lines.  If you want to make that a requirement of all the parsers, you could use data.lastIndexOf('\n') and get better performance.  Ideally, the buffer will end on a newline character.

Reply:
Okay, so I think the parser should be responsible for just a single record. So I've moved the line handling into TcpLineSource (as the name would suggest) and removed it from the GraphitePlaintextToRecordParser (it was never in the Telegraf one). 

This will also lead into the discussion about partial failure. That failure will not occur in the parser anymore but in the source. So I think this leads to more consistent semantics where the parser should always throw and the source can skip individual records without discarding everything. 

There are still questions of soft failure, such as with tags in Graphite, and what should be done with records that could be partially recovered. I expect we'll need to address those case by case.